Check input files exist before submitting to SLURM

PerilousApricot opened the pull request:
For some reason, condor-ce and BLAH can get out of sync -- the net
result is that BLAH is told to submit jobs whose input files (and even
parent directories!) don't exist in condor's spol directory. These jobs
are doomed to fail and be later restarted, causing a lot of unnecessary
churn on the SLURM scheduler.

Before submitting jobs to SLURM, verify that the input files exist. If
they don't, error out early.

PerilousApricot committed:
Check input files exist before submitting to SLURM

For some reason, condor-ce and BLAH can get out of sync -- the net
result is that BLAH is told to submit jobs whose input files (and even
parent directories!) don't exist in condor's spol directory. These jobs
are doomed to fail and be later restarted, causing a lot of unnecessary
churn on the SLURM scheduler.

Before submitting jobs to SLURM, verify that the input files exist. If
they don't, error out early.

brianhlin commented:
@PerilousApricot what version of condor, htcondor-ce, and the blahp are you running? That's definitely strange behavior. At a glance this looks pretty good but I'd also be interested in finding the root cause of this behavior. What indication do you have that the blahp is submitting jobs with files that aren't in the spool dir?

brianhlin commented:
@edquist could you take a look at this

edquist commented on src/scripts/blah_common_submit_functions.sh:
```diff
@@ -130,6 +130,34 @@ function bls_fl_subst_and_accumulate ()
+
+  local ind
+  bls_fl_test_exists_result=
+  for (( ind=0 ; ind < $last_argument ; ind++ )) ; do
```
Very minor style thing here, but you can skip the `$` for `$last_argument` in the `(( ))` expression, as is done for `ind`.  (But I'm guessing the intention is just to match similar existing code.)

edquist commented on src/scripts/blah_common_submit_functions.sh:
```diff
@@ -130,6 +130,34 @@ function bls_fl_subst_and_accumulate ()
+  bls_fl_test_exists_result=
+  for (( ind=0 ; ind < $last_argument ; ind++ )) ; do
+      bls_fl_subst $container_name $ind "@@F_LOCAL"
+      if [ ! -z "$bls_fl_subst_result" -a ! -f "$bls_fl_subst_result" ] ; then
```
I'm not super familiar with this, but for the `! -z "$bls_fl_subst_result"` part of this condition -- under what circumstances can `$bls_fl_subst_result` be empty?  My read of `bls_fl_subst` and `bls_fl_add_value` is that `local_file_name`/`remote_file_name` always get specified together, and an empty `local_file_name` is not allowed in `bls_fl_add_value`.

I'm not really objecting, because `bls_fl_subst`, which you're calling, makes it seem like it will return an empty `$bls_fl_subst_result` if `$f_local` or `$f_remote` is empty, it's just odd to me because it looks like that condition (either being empty) can't ever happen.  Ah well.

edquist reviewed (approved):
Looks fine & functional to me (a BLAH newbie).

Added some comments, but nothing needs to be changed unless you want to.

edquist commented:
PS: The BLAH is so endearing!

edquist commented:
Oh look at that:

> At least one approved review is required by reviewers with write access.

I am not that, which is why, despite me having reviewed and approved these changes, a review is still required and merging is blocked.

PerilousApricot commented:
Hi @edquist I have no problem with making your suggestions - I basically cargo-culted `bls_fl_subst_and_dump` because it had the "iterate over the list" construct I needed. I chose the (godawful, IMO) formatting to be consistent with what's done elsewhere.

Lemme know and I can do a rebase for you.

@brianhlin I think what might be happening (which is why you're not seeing things elsewhere) is SLURM is timing out to submit the jobs, then the source/destination jobs in condor desync. Condor correctly handles the "top half" and removes the spool directory, but the "bottom" half keeps getting resubmitted.

```
[root@ce3 ~]# rpm -q condor
condor-8.6.8-1.osg34.el7.x86_64
[root@ce3 ~]# rpm -q htcondor-ce
htcondor-ce-3.0.4-1.osg34.el7.noarch
[root@ce3 ~]# rpm -q blahp
blahp-1.18.35.bosco-1.osg34.el7.x86_64
```

PerilousApricot commented:
```
[root@ce3 ~]# condor_ce_q -constraint 'JobStatus=?=5' -long | grep '^HoldReason ' | sed 's,exist:.*,exist:,' | sort | uniq -c
    170 HoldReason = "Attempts to submit failed: "
     37 HoldReason = "no jobId in submission script's output (stdout:) (stderr: <blah> execute_cmd: 30 seconds timeout expired, killing child process.-)"
     36 HoldReason = "submission command failed (exit code = -15) (stdout:) (stderr: <blah> execute_cmd: 30 seconds timeout expired, killing child process.- <blah> killed by signal  15.-)"
    144 HoldReason = "submission command failed (exit code = 1) (stdout:Error-) (stderr:Error: Workdir doesn't exist-)"
```

edquist commented:
> I chose the (godawful, IMO) formatting to be consistent with what's done elsewhere.

@PerilousApricot, yeah it makes sense at least to keep it consistent with what else is there.  I'm not a BLAH maintainer and i don't know how they feel about such stylistic things.  Maybe if anybody cares they can clean up all the awkward bits in a separate PR.

Regardless, somebody with write access to this repo (ie, somebody other than myself) must approve this PR before merging is allowed, apparently.

PerilousApricot committed:
Check workdir exists for jobs submitted to SLURM

For some reason, condor-ce is pulling the rug out of jobs in SLURM and
deleting the spool directory. This is annoying because SLURM spends a lot
of time scheduling jobs that will never work. Fail early if we detect
this.

PerilousApricot commented:
BTW - does BLAH have a magic error code I could pass back to tell condor-ce "this is a terminal failure, don't retry"

PerilousApricot commented:
I think I've found a symptom?
```
/var/log/condor-ce/JobRouterLog:02/07/18 12:42:01 JobRouter (src=30034.0,dest=30036.0): removing orphaned destination job with no matching source job.
<snip, repeats every minute or so>
/var/log/condor-ce/JobRouterLog:02/07/18 20:29:35 JobRouter (src=30034.0,dest=30036.0): removing orphaned destination job with no matching source job.
/var/log/condor-ce/JobRouterLog:02/07/18 20:30:28 JobRouter (src=30034.0,dest=30036.0): removing orphaned destination job with no matching source job.
/var/log/condor-ce/JobRouterLog:02/07/18 20:31:21 JobRouter (src=30034.0,dest=30036.0): removing orphaned destination job with no matching source job.
```
but other logs claims
```
/var/log/condor-ce/SchedLog:02/07/18 12:27:59 Submitting new job 30034.0
/var/log/condor-ce/SchedLog:02/07/18 12:28:07 Transferring files for jobs 30034.0
/var/log/condor-ce/SchedLog:GlobalJobId = "ce3.accre.vanderbilt.edu#30034.0#1518028080"
/var/log/condor-ce/SchedLog:02/07/18 12:28:10 Received proxy for job 30034.0
/var/log/condor-ce/SchedLog:02/07/18 12:28:11 Job 30034.0 released from hold: Data files spooled
/var/log/condor-ce/SchedLog:02/07/18 12:31:09 Remove jobs 30034.0
/var/log/condor-ce/SchedLog:02/07/18 12:31:09 Finished Remove jobs 30034.0
/var/log/condor-ce/AuditLog:02/07/18 12:27:59 (cid:8937094) Submitting new job 30034.0
/var/log/condor-ce/AuditLog:02/07/18 12:28:07 (cid:8937103) Transferring files for jobs 30034.0
/var/log/condor-ce/AuditLog:02/07/18 12:28:10 (cid:8937103) Received proxy for job 30034.0
/var/log/condor-ce/AuditLog:02/07/18 12:31:09 (cid:8938185) Remove jobs 30034.0
/var/log/condor-ce/AuditLog:02/07/18 12:31:09 (cid:8938185) Finished Remove jobs 30034.0
```

brianhlin commented:
Hey @PerilousApricot, we're just waiting on a review from someone on the condor team. As for the root cause of the problem, could you show us an instance where you see those `JobRouterLog` messages and lines in the `GridmanagerLog` corresponding to the `dest` job?

brianhlin reviewed (approved):
Jaime gave his blessing. Merging

PerilousApricot commented:
Where should we follow up on whatever s happening?

On Fri, Feb 23, 2018 at 10:53 Brian Lin <notifications@github.com> wrote:

> Merged #67 <https://github.com/osg-bosco/BLAH/pull/67>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/osg-bosco/BLAH/pull/67#event-1489490065>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAFsqnmmEkjz6NgC24hYCzORO-30ZRFUks5tXu0HgaJpZM4R8Job>
> .
>
-- 
It's dark in this basement.

brianhlin commented:
Ticket here: https://jira.opensciencegrid.org/browse/SOFTWARE-3154

Targeted for the next release. Did you see my previous question about tracking down the root cause?

> As for the root cause of the problem, could you show us an instance where you see those JobRouterLog messages and lines in the GridmanagerLog corresponding to the dest job?

PerilousApricot commented:
Ah sorry, missed that comment (was on vacation). I'll follow up on the
ticket.

brianhlin commented:
Do you have an OSG JIRA account? If you don't we'll have to create one for you

PerilousApricot commented:
I don't appear to...

brianhlin commented:
Could you hop on OSG's slack (http://opensciencegrid.slack.com/)? We can sort out the details there

PerilousApricot commented:
I don't have a login there either ... Can you make me one? I forgot how
slack does it

On Fri, Feb 23, 2018 at 12:06 PM, Brian Lin <notifications@github.com>
wrote:

> Could you hop on OSG's slack (http://opensciencegrid.slack.com/)? We can
> sort out the details there
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/osg-bosco/BLAH/pull/67#issuecomment-368090909>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFsqg-rtPxrsBy81nzmzXwWluLLDpzHks5tXv4sgaJpZM4R8Job>
> .
>



-- 
--
Andrew Melo

brianhlin commented:
You should just be able to make your own account on slack, it shouldn't be limited like our JIRA is

PerilousApricot commented:
I get the following: "Don't have an @opensciencegrid.org, @osgconnect.net, @
iu.edu, @uchicago.edu, @unl.edu, or @wisc.edu email address?
Contact your Workspace Administrator for an invitation."

brianhlin commented:
I sent a follow-up to your vanderbilt email
